Adding Multiindex support to dataframe pivot function(Fixes #21425)

NikhilKumarM opened the pull request:
- [x] closes #21425
- [x] tests added/passed:
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew : Now dataframe.pivot function can take multiple columns as index 
        Added extra case to handle indexing on multiple colums in dataframe.pivot function

NikhilKumarM committed:
Adding Multiindex support to dataframe pivot function(Fixes #21425)

gfyoung commented:
@NikhilKumarM : Off to a good start here!  Going to need a `whatsnew` entry + test (can use the example that @jorisvandenbossche gave).

gfyoung commented:
@NikhilKumarM : You actually need to add the `whatsnew` entry (search for the `v0.23.2.txt` file and test cases to your PR changes 😄 .

jorisvandenbossche commented:
@NikhilKumarM Did you push your new changes? Because they do not show up on github here. Normally you only need to add new commits to the same branch, push that to github, and the PR here should get updated.

gfyoung commented:
@jorisvandenbossche : If you look at the PR description, you can see that @NikhilKumarM accidentally added them there.  If this PR goes unresponsive, I can bring this to the finish line.

NikhilKumarM committed:
added whatsnew entry in v0.23.2.txt file

NikhilKumarM committed:
Added test case for testing dataframe pivot function with multile columns as index

pep8speaks commented:
Hello @NikhilKumarM! Thanks for updating the PR.



 - In the file [`pandas/tests/reshape/test_pivot.py`](https://github.com/pandas-dev/pandas/blob/100a4bcc2c34196c4039218d33a857b903d88ed1/pandas/tests/reshape/test_pivot.py), following are the PEP8 issues :

> [Line 292:80](https://github.com/pandas-dev/pandas/blob/100a4bcc2c34196c4039218d33a857b903d88ed1/pandas/tests/reshape/test_pivot.py#L292): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 293:80](https://github.com/pandas-dev/pandas/blob/100a4bcc2c34196c4039218d33a857b903d88ed1/pandas/tests/reshape/test_pivot.py#L293): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)
> [Line 295:80](https://github.com/pandas-dev/pandas/blob/100a4bcc2c34196c4039218d33a857b903d88ed1/pandas/tests/reshape/test_pivot.py#L295): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 296:37](https://github.com/pandas-dev/pandas/blob/100a4bcc2c34196c4039218d33a857b903d88ed1/pandas/tests/reshape/test_pivot.py#L296): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','

gfyoung reviewed (approved):
Nice!  LGTM!

cc @jorisvandenbossche

NikhilKumarM committed:
Made changes as per PEP8 requirements

NikhilKumarM committed:
changed the type of exp_columns to Index

jorisvandenbossche commented on doc/source/whatsnew/v0.23.2.txt:
```diff
@@ -66,7 +66,7 @@ Bug Fixes
 **Reshaping**
 
--
+- Bug in: DataFrame.pivot() function where error was raised when multiple columns are set as index
```
Can you move this to the `0.24.0.txt` file? And you can mention it in the Enhancements section

(as this is an new feature, not really a bug fix, I think, as it was just not yet implemented)

jorisvandenbossche commented on pandas/core/reshape/reshape.py:
```diff
@@ -392,12 +392,22 @@ def pivot(self, index=None, columns=None, values=None):
         cols = [columns] if index is None else [index, columns]
         append = index is None
         indexed = self.set_index(cols, append=append)
+    # adding the support for multi-index in pivot function
```
this comment is not really needed I think

jorisvandenbossche commented on pandas/tests/reshape/test_pivot.py:
```diff
@@ -283,6 +283,23 @@ def test_pivot_multi_functions(self):
         expected = concat([means, stds], keys=['mean', 'std'], axis=1)
         tm.assert_frame_equal(result, expected)
 
+    # adding the test case for multiple columns as index (#21425)
```
please move this comment to the first line of the test function

jorisvandenbossche commented on pandas/tests/reshape/test_pivot.py:
```diff
@@ -283,6 +283,23 @@ def test_pivot_multi_functions(self):
+        expected = DataFrame([[0, 1], [2, 3], [4, 5], [6, 7]],
+                             exp_index,
+                             exp_columns)
+        tm.assert_frame_equal(result, expected)
```
Can you add an example where the `values` argument is not mentioned? eg `df.pivot(index=['lev1', 'lev2'], columns='lev3')` 
(I think this will not yet work, but should work)

jorisvandenbossche reviewed (commented):
- Can you also add the ability to pass multiple values for `columns` ?
- Can you also update the docstring of `DataFrame.pivot` ?

NikhilKumarM committed:
Moved whatsnew entry from v0.23.2.txt to v0.24.0.txt

NikhilKumarM committed:
Removed the comments that are not required

NikhilKumarM commented:
I have not worked on docstrings before. Is there any good resource to start doing that? @jorisvandenbossche

WillAyd commented:
@NikhilKumarM there is a section in the contributing guide focused on docstrings:

https://pandas.pydata.org/pandas-docs/stable/contributing_docstring.html

NikhilKumarM commented on pandas/core/reshape/reshape.py:
```diff
@@ -392,12 +392,21 @@ def pivot(self, index=None, columns=None, values=None):
+            index = MultiIndex.from_arrays([index, self[columns]])
+        # added this case to handle multi-index
+        elif isinstance(index, list):
+            indexes = []
```
sure!

NikhilKumarM commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -16,6 +16,8 @@ Other Enhancements
 - :func:`Series.mode` and :func:`DataFrame.mode` now support the ``dropna`` parameter which can be used to specify whether NaN/NaT values should be considered (:issue:`17534`)
 - :func:`to_csv` now supports ``compression`` keyword when a file handle is passed. (:issue:`21227`)
 - :meth:`Index.droplevel` is now implemented also for flat indexes, for compatibility with MultiIndex (:issue:`21115`)
+- :func:'Dataframe.pivot' now supports multiple columns as index. (:issue:'21425')
```
Sure. I will make those changes.

NikhilKumarM replied:
corrected

NikhilKumarM committed:
corrected whatsnew entry

NikhilKumarM committed:
Changed indexes to list comprehension

NikhilKumarM commented on pandas/core/reshape/reshape.py:
```diff
@@ -392,12 +392,21 @@ def pivot(self, index=None, columns=None, values=None):
         if index is None:
             index = self.index
+            index = MultiIndex.from_arrays([index, self[columns]])
+        # added this case to handle multi-index
```
Done

NikhilKumarM commented:
Why is continuous integration failing ?

gfyoung commented:
> Why is continuous-integration/travis-ci failing?

@NikhilKumarM : Restarted Travis for you.  Failure unrelated to PR.

jreback commented on pandas/core/reshape/reshape.py:
```diff
@@ -392,12 +392,19 @@ def pivot(self, index=None, columns=None, values=None):
         if index is None:
             index = self.index
+            index = MultiIndex.from_arrays([index, self[columns]])
+        elif isinstance(index, list):
```
use ``is_list_like`` instead here

NikhilKumarM replied:
Sure. I will do it.

jreback reviewed (changes requested):
small change otherwise lgtm. does the doc string need updating?

NikhilKumarM committed:
used is_list_like to check if input is list

NikhilKumarM committed:
resolving merge conflict in whatsnew

NikhilKumarM committed:
trying to solve merge conflict

NikhilKumarM committed:
Merge branch 'master' into master

NikhilKumarM commented:
@jreback @gfyoung @jorisvandenbossche  
Could you please check my last commit and suggest changes. I have modified the code to handle when there is no 'values' argument for 'pivot' function and also added a test case for that. But I do not know if have done it in an optimal way.

NikhilKumarM committed:
Handling the case when there is no values argument in pivot func

NikhilKumarM committed:
added one more test case for pivot function with multi column index

gfyoung commented:
@NikhilKumarM : Got a linting error on Travis:

https://travis-ci.org/pandas-dev/pandas/jobs/398632775#L2904

NikhilKumarM commented:
Fixed the linting error.

NikhilKumarM committed:
added whitespace to remove linting error

NikhilKumarM commented:
@gfyoung @jreback  Could you please give me your feedback on my last commit?

jreback commented on pandas/core/reshape/reshape.py:
```diff
@@ -395,15 +395,29 @@ def pivot(self, index=None, columns=None, values=None):
     if values is None:
-        cols = [columns] if index is None else [index, columns]
+        if index is None:
+            cols = [columns]
```
do we have tests for each one of these paths here? e.g. branching a lot more now

NikhilKumarM replied:
Yes, this is an enhancement. I have added the test cases and need to update the docs. Thanks for your time.

jreback reviewed (commented):
pls add a whatsnew note for 0.24. I guess this is a bug fix, though was this documented before? if so its an enhancement. Can you also add an exampl ein reshape.rst with this. And in the pivot doc-string as well.

jreback commented:
also needs to rebase

jreback commented:
can you merge master

NikhilKumarM commented:
Sure!

jreback commented on pandas/core/reshape/reshape.py:
```diff
@@ -395,15 +395,29 @@ def pivot(self, index=None, columns=None, values=None):
-        cols = [columns] if index is None else [index, columns]
+        if index is None:
+            cols = [columns]
+        else:
```
can you add a comment explaining what is going on here, ideally we would have a comment for each branch of the if/else

jreback reviewed (changes requested):
can you merge master & add a whatsnew note.

jreback commented:
can you update

jreback commented:
can you merge master

NikhilKumarM commented:
I think pivot function has been removed from reshape.py and put in pivot.py. I will rebase with the master and make necessary changes by this weekend. Thank you !!

jreback commented:
can you merge master and update

WillAyd commented:
Closing as stale - ping if you'd like to reopen and continue working this
